modules/ROOT/edits/dataweave-type-system.adoc

```diff
@@ -0,0 +1,358 @@
+
+So, if you call the function `isNumeric` with `null` (`isNumeric(null)`), the function does not throw an exception. At runtime, the function that is dispatching algorithm selects the correct function based on the types of the values you use. The function returns `false` because of the definition of `isNumeric` for `null` as the argument.
+
+== Composite Types
+
+A composite types contains other values. The composite types are `Array`, `Object` and `Function`.
+
+=== Array Type
+
+An  array is of the type `Array&lt;T&gt;` in which `T` is a type parameter that defines the type of the elements inside the array. For example, the syntax to define a variable with the type `Array` of `Number` is:
+
+----
+var idsList: Array<Number> = [1, 22, 333, 4444]
+----
+
+The variable `idsList` only passes the type-checking phase when the array that is assigned to it only contains values of the `Number` type.
+
+=== Object Type
+
+You can define an `Object` type in two ways: `Object` or `{}` both define an open object and don't specify any constraint to their key-value pairs. See also, <<close_open_objects>>.
+
+The syntax to define an `Object` type with a set of constraints for its key-value pairs is very similar to defining a value of an `Object` type:
+
+----
+{
+    keyName @(attrName: AttrType): valueType,
+    ...
+}
+----
+
+For example, to define a `User` type with `firstName` `String`, `lastName` `String `and `age` `Number`, the syntax is:
+
+.DataWeave Script
+----
+%dw 2.0
+output application/json
+
+type User = {
+    firstName: String,
+    lastName: String,
+    age: Number
+}
+----
+
+You can apply some modifiers to each key-value pair to determine if a field is always present or if the field can be repeated.
+
+For repeated fields use `&#42;`. For example, the following type allows you to have multiple `lastName` fields:
+
+.DataWeave Script
+----
+%dw 2.0
+output application/json
+
+{
+    firstName: String,
+    lastName*: String,
+    age: Number
+}
+----
+
+For conditional fields, use `?`. For example, the following type allows the age field to be present or not:
+
+.DataWeave Script
+----
+%dw 2.0
+output application/json
+
+{
+    firstName: String,
+    lastName*: String,
+    age?: Number
+}
+----
+
+[[close_open_objects]]
+==== Close and Open Objects
+
+A closed object means that the type only accepts an object value if there are no additional fields in the object,
+just those key-value pairs specified by the type.
+
+An open object only put constraints on the fields that are declared in the type, and if they are all verified, then the type accepts the value. The type accepts objects that have other fields apart of the ones mentioned in the declaration. All object types are open unless specified.
```

Comment:
I tried to guess what the previous paragraph means, but I wonder if the original text might use "type" or "object" when meaning the other? Please reread the text.
```suggestion
An _open_ object constrains only the fields that are declared in the type, although the type accepts objects that have fields not included in the declaration. All object types are open unless specified.
```
I also wonder if, since all objects are open unless specifically closed, we should start this topic rather with text something like
`Object` type objects are by default open, but they can be closed.
An _open_ object type <is this....>
A _closed_ object type <is that....>

Reply:
Please review GLOBALLY for proper distinction between the `Object` type and an object.